xtext/org.icyphy.linguafranca/src/org/icyphy/generator/GeneratorBase.xtend

```
@@ -373,24 +376,19 @@ abstract class GeneratorBase extends AbstractLinguaFrancaValidator {
     protected String targetName
     
     /**
-     * The level of logging or null if not given.
+     * The level of logging or null if not given. The default is ERROR.
```

Is there a particular reason to choose `ERROR` as default here? So far, the C++ target used the `INFO` level as default, which to me personally seems to be the more reasonable default. Maybe we could also leave this as `null` if not specified and let the targets decide what the default is

---

I agree. The default should be INFO.

---

I was under the assumption that this was modeled after the reactor-ts implementation -- that is where the levels originate from. I would assume that the default is to display only errors (the lowest logging level). Up from `ERROR`, the subsequent levels add more detail. In order of verbosity: `WARN`, `INFO`, `LOG`, `DEBUG`. I'm fine revisiting any aspect of this design, but I'd definitely suggest we implement the same thing across targets.

---

Well, where this originates from is debatable :wink: Logging levels were supported in C++ long before it was a LF target. But I don't want to make this about who was first. By default, the logging level in C++ it set to `INFO` because this logging level shows relevant information that might be useful to the programmer. This covers a note at start and end of execution, similar as it is displayed in C. To me its weird that `ERROR` should be default and I actively have to increase the level in order to see warnings. If `INFO` is default, then when you start using LF and write your first program you the relevant information displayed alongside your program. If you don't want this information you switch to `WARNING`. And if that annoys you still, you can ignore any warnings by switching to `ERROR` but then you are on your own. Similar, if something goes wrong, you can switch to `LOG` or `DEBUG` in order to see more messages. In any case, when you don't specify anything you see all information that is considered useful for the average user.

---

It's the specific logging levels that I was referring to that originate from reactor-ts. I also didn't come up with them, I borrowed them from the logging package we use.

Your argument for having `INFO` as the default level sounds reasonable to me. C currently supports only two log levels (`INFO` and `DEBUG`) and has an orthogonal `VERBOSE` flag. I think we should eliminate `VERBOSE` and add support for the missing levels, bringing the C and C++ targets in compliance with each other.

---

OK, this works for me, although it seems odd to have "superfluous" levels. We could also just eliminate some...

---

In the C target documentation now, we are encouraging application programmers to use info_print() rather than printf() because this can be intercepted and sent to an ncurses window, or in an embedded platform, to an LCD screen or some other output.  I don't think that suppressing warnings and info by default is a good idea.

Also, in the C target, VERBOSE is not a user-facing thing at all. It's just an implementation detail. A better implementation would not have it at all. It is there for legacy reasons.

---

I agree that having `INFO` by default is good. My question is whether or not we should eliminate any of the levels. C and C++ only use a subset. I think it is preferable that all targets use the same set of levels. We can either add support for extra levels in C/C++ or remove some in TS.

---

I don't understand this proposal. C has five levels: error, warning, info, log, debug.  Currently, the API only allows you to choose between info (the default), log, and debug. But this would be easy to extend, but it is hardly a priority, since I have no reason to want to mask errors or warnings.  I suppose that when we have a reason, we can add that capability.  If you remove the levels now, then no such extension will be possible in the future without rewriting all the source code that generates the messages.

---

OK, if it's not a priority then let's leave it as-is.